File: app/src/main/java/com/pkmnapps/activitydo/adapters/ActivityContentAdapter.java
package com.pkmnapps.activitydo.adapters;

import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.pkmnapps.activitydo.ImageViewFullscreen;
import com.pkmnapps.activitydo.MConstants;
import com.pkmnapps.activitydo.R;
import com.pkmnapps.activitydo.custominterfaces.TaskActivityInterface;
import com.pkmnapps.activitydo.databasehelpers.DBHelperListItems;
import com.pkmnapps.activitydo.dataclasses.ImageWidget;
import com.pkmnapps.activitydo.dataclasses.ListWidget;
import com.pkmnapps.activitydo.dataclasses.SimpleTextWidget;
import com.pkmnapps.activitydo.dataclasses.Widget;

import java.util.List;

public class ActivityContentAdapter extends RecyclerView.Adapter{

    List<Widget> widgets;
    TaskActivityInterface taskActivityInterface;
    public class MyTextViewHolder extends RecyclerView.ViewHolder {
        public TextView head,body;
        public ImageButton more;
        public MyTextViewHolder(View view) {
            super(view);
            head = (TextView)view.findViewById(R.id.head_textView);
            body = (TextView)view.findViewById(R.id.body_textView);
            more = (ImageButton)view.findViewById(R.id.more_button);
        }
    }
    public class MyListViewHolder extends RecyclerView.ViewHolder {
        public TextView head;
        public RecyclerView recyclerView;
        public ImageButton more;

        public MyListViewHolder(View view) {
            super(view);
            head = (TextView)view.findViewById(R.id.head_textView);
            recyclerView = (RecyclerView)view.findViewById(R.id.recycler_view);
            more = (ImageButton)view.findViewById(R.id.more_button);

        }
    }
    public class MyImageViewHolder extends RecyclerView.ViewHolder {
        public ImageView imageView;
        public ImageButton more;

        public MyImageViewHolder(View view) {
            super(view);
            imageView = (ImageView)view.findViewById(R.id.widget_imageView);
            more = (ImageButton)view.findViewById(R.id.more_button);

        }
    }
    public class MyAudioViewHolder extends RecyclerView.ViewHolder {

        public MyAudioViewHolder(View view) {
            super(view);

        }
    }

    public ActivityContentAdapter(List<Widget> widgets, TaskActivityInterface taskActivityInterface) {
        this.widgets = widgets;
        this.taskActivityInterface = taskActivityInterface;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v;
        switch (viewType){
            case MConstants.textW:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.simple_note_view, parent, false);
                return new MyTextViewHolder(v);
            case MConstants.listW:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.recycler_list_in_task_view, parent, false);
                return new MyListViewHolder(v);
            case MConstants.imageW:
                v = LayoutInflater.from(parent.getContext()).inflate(R.layout.image_widget_view, parent, false);
                return new MyImageViewHolder(v);
            case MConstants.audioW:
                return null;
            default:
                return null;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, final int position) {
        int type = getItemViewType(position);
        final Object widget = widgets.get(position).getObject();
        if(widget!=null && type!=0){
            switch (type){
                case MConstants.textW:
                    final SimpleTextWidget t = (SimpleTextWidget)widget;
                    ((MyTextViewHolder)holder).head.setText(t.getHead());
                    ((MyTextViewHolder)holder).body.setText(t.getBody());
                    ((MyTextViewHolder)holder).itemView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            taskActivityInterface.editWidget(widgets.get(holder.getAdapterPosition()));
                        }
                    });
                    ((MyTextViewHolder)holder).more.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            PopupMenu popupMenu = new PopupMenu(holder.itemView.getContext(),((MyTextViewHolder)holder).more);

                            popupMenu.getMenuInflater().inflate(R.menu.recycler_task_menu,popupMenu.getMenu());
                            popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                                @Override
                                public boolean onMenuItemClick(MenuItem item) {
                                    if(item.getItemId()==R.id.action_delete)
                                        taskActivityInterface.deleteWidget(widgets.get(holder.getAdapterPosition()));
                                    else if(item.getItemId()==R.id.action_change_activity){
                                        taskActivityInterface.changeActivtyOfWidget(MConstants.textW,t.getUid(),holder.getAdapterPosition());
                                    }
                                    return true;
                                }
                            });
                            popupMenu.show();
                        }
                    });
                    break;
                case MConstants.listW:
                    final ListWidget l = (ListWidget)widget;
                    ((MyListViewHolder)holder).head.setText(l.getHead());
                    RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(((MyListViewHolder)holder).itemView.getContext(),LinearLayoutManager.VERTICAL,false);
                    ((MyListViewHolder)holder).recyclerView.setLayoutManager(mLayoutManager);
                    ((MyListViewHolder)holder).recyclerView.setItemAnimator(new DefaultItemAnimator());
                    ((MyListViewHolder)holder).recyclerView.setNestedScrollingEnabled(false);
                            ((MyListViewHolder)holder).recyclerView
                            .setAdapter(new InViewListAdapter(new DBHelperListItems(((MyListViewHolder)holder).itemView.getContext())
                                    .get8ListItemsAsList(l.getUid())));
                    ((MyListViewHolder)holder).recyclerView.setLayoutFrozen(true);

                    ((MyListViewHolder)holder).itemView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            taskActivityInterface.editWidget(widgets.get(holder.getAdapterPosition()));
                        }
                    });
                    ((MyListViewHolder)holder).more.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            PopupMenu popupMenu = new PopupMenu(holder.itemView.getContext(),((MyListViewHolder)holder).more);

                            popupMenu.getMenuInflater().inflate(R.menu.recycler_task_menu,popupMenu.getMenu());
                            popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                                @Override
                                public boolean onMenuItemClick(MenuItem item) {
                                    if(item.getItemId()==R.id.action_delete)
                                        taskActivityInterface.deleteWidget(widgets.get(holder.getAdapterPosition()));
                                    else if(item.getItemId()==R.id.action_change_activity){
                                        taskActivityInterface.changeActivtyOfWidget(MConstants.listW,l.getUid(),holder.getAdapterPosition());
                                    }
                                    return true;
                                }
                            });
                            popupMenu.show();
                        }
                    });
                    break;
                case MConstants.imageW:
                    final ImageWidget i = (ImageWidget)widget;

                    Glide.with(((MyImageViewHolder)holder).imageView.getContext())
                            .load(Uri.parse(i.getImageUri()))
                            .into(((MyImageViewHolder)holder).imageView);
                    ((MyImageViewHolder)holder).itemView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            //open it in fullscreen
                            Intent intent = new Intent(((MyImageViewHolder)holder).itemView.getContext(), ImageViewFullscreen.class);
                            intent.putExtra("image", i.getImageUri());
                            ((MyImageViewHolder)holder).itemView.getContext().startActivity(intent);
                        }
                    });
                    ((MyImageViewHolder)holder).more.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            PopupMenu popupMenu = new PopupMenu(holder.itemView.getContext(),((MyImageViewHolder)holder).more);

                            popupMenu.getMenuInflater().inflate(R.menu.recycler_task_menu,popupMenu.getMenu());
                            popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                                @Override
                                public boolean onMenuItemClick(MenuItem item) {
                                    if(item.getItemId()==R.id.action_delete)
                                        taskActivityInterface.deleteWidget(widgets.get(holder.getAdapterPosition()));
                                    else if(item.getItemId()==R.id.action_change_activity){
                                        taskActivityInterface.changeActivtyOfWidget(MConstants.imageW,i.getUid(),holder.getAdapterPosition());
                                    }
                                    return true;
                                }
                            });
                            popupMenu.show();
                        }
                    });
                    break;
                case MConstants.audioW:
                    break;
            }
        }
    }

    @Override
    public int getItemCount() {
        return widgets.size();
    }

    @Override
    public int getItemViewType(int position) {
        return widgets.get(position).getType();
    }


}

File: app/src/main/java/com/pkmnapps/activitydo/adapters/ActivityContentAdapter.kt
package com.pkmnapps.activitydo.adapters

import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.PopupMenu
import android.widget.TextView
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.pkmnapps.activitydo.ImageViewFullscreen
import com.pkmnapps.activitydo.MConstants
import com.pkmnapps.activitydo.R
import com.pkmnapps.activitydo.custominterfaces.TaskActivityInterface
import com.pkmnapps.activitydo.databasehelpers.DBHelperListItems
import com.pkmnapps.activitydo.dataclasses.ImageWidget
import com.pkmnapps.activitydo.dataclasses.ListWidget
import com.pkmnapps.activitydo.dataclasses.SimpleTextWidget
import com.pkmnapps.activitydo.dataclasses.Widget
import org.junit.runner.RunWith

class ActivityContentAdapter(val widgets: MutableList<Widget?>?, val taskActivityInterface: TaskActivityInterface?) : RecyclerView.Adapter<Any?>() {
    class MyTextViewHolder(view: View?) : RecyclerView.ViewHolder(view) {
        val head: TextView?
        val body: TextView?
        val more: ImageButton?

        init {
            head = view.findViewById(R.id.head_textView)
            body = view.findViewById(R.id.body_textView)
            more = view.findViewById(R.id.more_button)
        }
    }

    class MyListViewHolder(view: View?) : RecyclerView.ViewHolder(view) {
        val head: TextView?
        val recyclerView: RecyclerView?
        val more: ImageButton?

        init {
            head = view.findViewById(R.id.head_textView)
            recyclerView = view.findViewById(R.id.recycler_view)
            more = view.findViewById(R.id.more_button)
        }
    }

    class MyImageViewHolder(view: View?) : RecyclerView.ViewHolder(view) {
        val imageView: ImageView?
        val more: ImageButton?

        init {
            imageView = view.findViewById(R.id.widget_imageView)
            more = view.findViewById(R.id.more_button)
        }
    }

    class MyAudioViewHolder(view: View?) : RecyclerView.ViewHolder(view)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val v: View?
        return when (viewType) {
            MConstants.textW -> {
                v = LayoutInflater.from(parent.context).inflate(R.layout.simple_note_view, parent, false)
                MyTextViewHolder(v)
            }
            MConstants.listW -> {
                v = LayoutInflater.from(parent.context).inflate(R.layout.recycler_list_in_task_view, parent, false)
                MyListViewHolder(v)
            }
            MConstants.imageW -> {
                v = LayoutInflater.from(parent.context).inflate(R.layout.image_widget_view, parent, false)
                MyImageViewHolder(v)
            }
            MConstants.audioW -> null
            else -> null
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val type = getItemViewType(position)
        val widget = widgets.get(position).getObject()
        if (widget != null && type != 0) {
            when (type) {
                MConstants.textW -> {
                    val t = widget as SimpleTextWidget
                    (holder as MyTextViewHolder).head.setText(t.head)
                    (holder as MyTextViewHolder).body.setText(t.body)
                    (holder as MyTextViewHolder).itemView.setOnClickListener { taskActivityInterface.editWidget(widgets.get(holder.getAdapterPosition())) }
                    (holder as MyTextViewHolder).more.setOnClickListener(View.OnClickListener {
                        val popupMenu = PopupMenu(holder.itemView.context, (holder as MyTextViewHolder).more)
                        popupMenu.menuInflater.inflate(R.menu.recycler_task_menu, popupMenu.menu)
                        popupMenu.setOnMenuItemClickListener { item ->
                            if (item.itemId == R.id.action_delete) taskActivityInterface.deleteWidget(widgets.get(holder.getAdapterPosition())) else if (item.itemId == R.id.action_change_activity) {
                                taskActivityInterface.changeActivtyOfWidget(MConstants.textW, t.uid, holder.getAdapterPosition())
                            }
                            true
                        }
                        popupMenu.show()
                    })
                }
                MConstants.listW -> {
                    val l = widget as ListWidget
                    (holder as MyListViewHolder).head.setText(l.head)
                    val mLayoutManager: RecyclerView.LayoutManager = LinearLayoutManager((holder as MyListViewHolder).itemView.context, LinearLayoutManager.VERTICAL, false)
                    (holder as MyListViewHolder).recyclerView.setLayoutManager(mLayoutManager)
                    (holder as MyListViewHolder).recyclerView.setItemAnimator(DefaultItemAnimator())
                    (holder as MyListViewHolder).recyclerView.setNestedScrollingEnabled(false)
                    (holder as MyListViewHolder).recyclerView
                            .setAdapter(InViewListAdapter(DBHelperListItems((holder as MyListViewHolder).itemView.context)
                                    .get8ListItemsAsList(l.uid)))
                    (holder as MyListViewHolder).recyclerView.suppressLayout(true)
                    (holder as MyListViewHolder).itemView.setOnClickListener { taskActivityInterface.editWidget(widgets.get(holder.getAdapterPosition())) }
                    (holder as MyListViewHolder).more.setOnClickListener(View.OnClickListener {
                        val popupMenu = PopupMenu(holder.itemView.context, (holder as MyListViewHolder).more)
                        popupMenu.menuInflater.inflate(R.menu.recycler_task_menu, popupMenu.menu)
                        popupMenu.setOnMenuItemClickListener { item ->
                            if (item.itemId == R.id.action_delete) taskActivityInterface.deleteWidget(widgets.get(holder.getAdapterPosition())) else if (item.itemId == R.id.action_change_activity) {
                                taskActivityInterface.changeActivtyOfWidget(MConstants.listW, l.uid, holder.getAdapterPosition())
                            }
                            true
                        }
                        popupMenu.show()
                    })
                }
                MConstants.imageW -> {
                    val i = widget as ImageWidget
                    Glide.with((holder as MyImageViewHolder).imageView.getContext())
                            .load(Uri.parse(i.imageUri))
                            .into((holder as MyImageViewHolder).imageView)
                    (holder as MyImageViewHolder).itemView.setOnClickListener { //open it in fullscreen
                        val intent = Intent((holder as MyImageViewHolder).itemView.context, ImageViewFullscreen::class.java)
                        intent.putExtra("image", i.imageUri)
                        (holder as MyImageViewHolder).itemView.context.startActivity(intent)
                    }
                    (holder as MyImageViewHolder).more.setOnClickListener(View.OnClickListener {
                        val popupMenu = PopupMenu(holder.itemView.context, (holder as MyImageViewHolder).more)
                        popupMenu.menuInflater.inflate(R.menu.recycler_task_menu, popupMenu.menu)
                        popupMenu.setOnMenuItemClickListener { item ->
                            if (item.itemId == R.id.action_delete) taskActivityInterface.deleteWidget(widgets.get(holder.getAdapterPosition())) else if (item.itemId == R.id.action_change_activity) {
                                taskActivityInterface.changeActivtyOfWidget(MConstants.imageW, i.uid, holder.getAdapterPosition())
                            }
                            true
                        }
                        popupMenu.show()
                    })
                }
                MConstants.audioW -> {
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return widgets.size
    }

    override fun getItemViewType(position: Int): Int {
        return widgets.get(position).getType()
    }
}